File: go/tracer.go
package tokenlens

import (
	"fmt"
	"math"
	"sort"
)

// UsageRecord is one LLM call: what it cost, how long it took, and who it was for.
type UsageRecord struct {
	Model        string
	InputTokens  int64
	OutputTokens int64
	LatencyMs    float64
	Feature      string
	Tenant       string
	Timestamp    float64 // epoch seconds; used by the rolling-window creep detector
}

// Cost rates the record against the default price book.
func (r UsageRecord) Cost() float64 { return CostOf(r.Model, r.InputTokens, r.OutputTokens) }

// TotalTokens is input plus output.
func (r UsageRecord) TotalTokens() int64 { return r.InputTokens + r.OutputTokens }

// DimensionStat accumulates usage for one value of a dimension.
type DimensionStat struct {
	Key          string
	Calls        int
	InputTokens  int64
	OutputTokens int64
	Cost         float64
	LatencySum   float64
}

// AvgLatencyMs is the mean latency across the dimension's calls.
func (s *DimensionStat) AvgLatencyMs() float64 {
	if s.Calls == 0 {
		return 0
	}
	return math.Round(s.LatencySum/float64(s.Calls)*100) / 100
}

// TotalTokens is input plus output.
func (s *DimensionStat) TotalTokens() int64 { return s.InputTokens + s.OutputTokens }

// Anomaly flags a dimension that is expensive relative to its peers right now.
type Anomaly struct {
	Key      string
	Metric   string
	Value    float64
	Baseline float64
	Factor   float64
}

// Creep flags a dimension whose cost rate is rising over time.
//
// The median detector only catches a dimension that is expensive relative to its
// peers right now. It misses one that is slowly growing but still cheaper than
// the noisy ones. Comparing a recent window against an earlier baseline window
// catches that trend.
type Creep struct {
	Key          string
	BaselineRate float64 // cost per second over the baseline window
	RecentRate   float64 // cost per second over the recent window
	Factor       float64
}

// Report is the aggregate view: spend by dimension, budget state, and anomalies.
type Report struct {
	TotalCost      float64
	TotalCalls     int
	ByDimension    map[string]*DimensionStat
	BudgetExceeded bool
	Anomalies      []Anomaly
}

func dimensionOf(r UsageRecord, dimension string) (string, error) {
	switch dimension {
	case "feature":
		return r.Feature, nil
	case "tenant":
		return r.Tenant, nil
	case "model":
		return r.Model, nil
	default:
		return "", fmt.Errorf("unknown dimension %q", dimension)
	}
}

// Aggregate rolls records up by the given dimension.
func Aggregate(records []UsageRecord, dimension string) (map[string]*DimensionStat, error) {
	stats := make(map[string]*DimensionStat)
	for _, r := range records {
		key, err := dimensionOf(r, dimension)
		if err != nil {
			return nil, err
		}
		s, ok := stats[key]
		if !ok {
			s = &DimensionStat{Key: key}
			stats[key] = s
		}
		s.Calls++
		s.InputTokens += r.InputTokens
		s.OutputTokens += r.OutputTokens
		s.Cost = Round6(s.Cost + r.Cost())
		s.LatencySum += r.LatencyMs
	}
	return stats, nil
}

// DetectAnomalies flags dimensions whose cost exceeds factor x the median dimension cost.
func DetectAnomalies(stats map[string]*DimensionStat, factor float64) []Anomaly {
	costs := make([]float64, 0, len(stats))
	for _, s := range stats {
		costs = append(costs, s.Cost)
	}
	if len(costs) < 3 {
		return nil
	}
	sort.Float64s(costs)
	median := costs[len(costs)/2]
	if median <= 0 {
		return nil
	}
	out := make([]Anomaly, 0)
	for _, s := range stats {
		if s.Cost > factor*median {
			out = append(out, Anomaly{
				Key:      s.Key,
				Metric:   "cost",
				Value:    s.Cost,
				Baseline: median,
				Factor:   math.Round(s.Cost/median*100) / 100,
			})
		}
	}
	sort.SliceStable(out, func(i, j int) bool { return out[i].Factor > out[j].Factor })
	return out
}

// DetectCreep flags dimensions whose cost rate rose from a baseline window to a recent one.
//
// Records with Timestamp < splitTime form the baseline window; the rest form the
// recent window. Cost is normalized by each window's duration to get a rate
// (cost/second), so uneven window lengths compare fairly. A dimension is flagged
// when its recent rate exceeds factor times its baseline rate.
//
// When splitTime is nil, the midpoint of the observed timestamp range is used.
func DetectCreep(records []UsageRecord, dimension string, splitTime *float64, factor float64) ([]Creep, error) {
	timed := make([]UsageRecord, 0, len(records))
	for _, r := range records {
		if r.Timestamp > 0 {
			timed = append(timed, r)
		}
	}
	if len(timed) < 2 {
		return nil, nil
	}

	lo, hi := timed[0].Timestamp, timed[0].Timestamp
	for _, r := range timed {
		lo = math.Min(lo, r.Timestamp)
		hi = math.Max(hi, r.Timestamp)
	}
	if hi == lo {
		return nil, nil
	}

	split := (lo + hi) / 2
	if splitTime != nil {
		split = *splitTime
	}
	baseDur := math.Max(split-lo, 1e-9)
	recentDur := math.Max(hi-split, 1e-9)

	baseCost := make(map[string]float64)
	recentCost := make(map[string]float64)
	for _, r := range timed {
		key, err := dimensionOf(r, dimension)
		if err != nil {
			return nil, err
		}
		if r.Timestamp < split {
			baseCost[key] += r.Cost()
		} else {
			recentCost[key] += r.Cost()
		}
	}

	out := make([]Creep, 0)
	for key, rc := range recentCost {
		baseRate := baseCost[key] / baseDur
		recentRate := rc / recentDur
		if baseRate <= 0 {
			continue // no baseline to compare against; new, not creep
		}
		if recentRate > factor*baseRate {
			out = append(out, Creep{
				Key:          key,
				BaselineRate: math.Round(baseRate*1e9) / 1e9,
				RecentRate:   math.Round(recentRate*1e9) / 1e9,
				Factor:       math.Round(recentRate/baseRate*100) / 100,
			})
		}
	}
	sort.SliceStable(out, func(i, j int) bool { return out[i].Factor > out[j].Factor })
	return out, nil
}

// BuildReport aggregates records and evaluates budget + anomalies in one pass.
// A nil budget means "no budget configured".
func BuildReport(records []UsageRecord, dimension string, budget *float64, anomalyFactor float64) (*Report, error) {
	stats, err := Aggregate(records, dimension)
	if err != nil {
		return nil, err
	}
	total := 0.0
	calls := 0
	for _, s := range stats {
		total += s.Cost
		calls += s.Calls
	}
	total = Round6(total)
	return &Report{
		TotalCost:      total,
		TotalCalls:     calls,
		ByDimension:    stats,
		BudgetExceeded: budget != nil && total > *budget,
		Anomalies:      DetectAnomalies(stats, anomalyFactor),
	}, nil
}

File: ts/src/tracer.ts
/**
 * Usage records and cost/latency aggregation with budget + anomaly detection.
 *
 * Core value: attribute LLM spend to a dimension (feature / tenant / model) and
 * surface budget breaches and cost anomalies - the numbers a FinOps/eng lead
 * asks for. Pure logic; no I/O.
 */

import { costOf, round6 } from "./pricing";

export type Dimension = "feature" | "tenant" | "model";

/** One LLM call: what it cost, how long it took, and who it was for. */
export interface UsageRecord {
  model: string;
  inputTokens: number;
  outputTokens: number;
  latencyMs: number;
  feature?: string;
  tenant?: string;
  /** epoch seconds; used by the rolling-window creep detector. 0 = untimed. */
  timestamp?: number;
}

/** Rates a record against the default price book. */
export function recordCost(r: UsageRecord): number {
  return costOf(r.model, r.inputTokens, r.outputTokens);
}

/** Input plus output tokens. */
export function recordTotalTokens(r: UsageRecord): number {
  return r.inputTokens + r.outputTokens;
}

/** Accumulated usage for one value of a dimension. */
export interface DimensionStat {
  key: string;
  calls: number;
  inputTokens: number;
  outputTokens: number;
  cost: number;
  latencySum: number;
}

/** The mean latency across a dimension's calls. */
export function avgLatencyMs(s: DimensionStat): number {
  return s.calls === 0 ? 0 : Math.round((s.latencySum / s.calls) * 100) / 100;
}

/** Input plus output tokens for a dimension. */
export function statTotalTokens(s: DimensionStat): number {
  return s.inputTokens + s.outputTokens;
}

/** A dimension that is expensive relative to its peers right now. */
export interface Anomaly {
  key: string;
  metric: string;
  value: number;
  baseline: number;
  factor: number;
}

/**
 * A dimension whose cost *rate* is rising over time (gradual creep).
 *
 * The plain median detector only catches a dimension that is expensive relative
 * to its peers right now. It misses one that is slowly growing but still
 * cheaper than the noisy ones. Comparing a recent window against an earlier
 * baseline window catches that trend.
 */
export interface Creep {
  key: string;
  /** cost per second over the baseline window */
  baselineRate: number;
  /** cost per second over the recent window */
  recentRate: number;
  factor: number;
}

/** Spend by dimension, budget state, and anomalies. */
export interface Report {
  totalCost: number;
  totalCalls: number;
  byDimension: Record<string, DimensionStat>;
  budgetExceeded: boolean;
  anomalies: Anomaly[];
}

function keyOf(r: UsageRecord, dimension: Dimension): string {
  switch (dimension) {
    case "feature":
      return r.feature ?? "unknown";
    case "tenant":
      return r.tenant ?? "unknown";
    case "model":
      return r.model;
    default:
      throw new Error(`unknown dimension '${dimension as string}'`);
  }
}

/** Rolls records up by the given dimension. */
export function aggregate(
  records: UsageRecord[],
  dimension: Dimension = "feature",
): Record<string, DimensionStat> {
  const stats: Record<string, DimensionStat> = {};
  for (const r of records) {
    const key = keyOf(r, dimension);
    let s = stats[key];
    if (!s) {
      s = { key, calls: 0, inputTokens: 0, outputTokens: 0, cost: 0, latencySum: 0 };
      stats[key] = s;
    }
    s.calls += 1;
    s.inputTokens += r.inputTokens;
    s.outputTokens += r.outputTokens;
    s.cost = round6(s.cost + recordCost(r));
    s.latencySum += r.latencyMs;
  }
  return stats;
}

/** Flags dimensions whose cost exceeds `factor` x the median dimension cost. */
export function detectAnomalies(
  stats: Record<string, DimensionStat>,
  factor = 3.0,
): Anomaly[] {
  const values = Object.values(stats);
  const costs = values.map((s) => s.cost).sort((a, b) => a - b);
  if (costs.length < 3) return [];
  const median = costs[Math.floor(costs.length / 2)];
  if (median <= 0) return [];
  return values
    .filter((s) => s.cost > factor * median)
    .map((s) => ({
      key: s.key,
      metric: "cost",
      value: s.cost,
      baseline: median,
      factor: Math.round((s.cost / median) * 100) / 100,
    }))
    .sort((a, b) => b.factor - a.factor);
}

/**
 * Flags dimensions whose cost *rate* rose from a baseline window to a recent one.
 *
 * Records with `timestamp < splitTime` form the baseline window; the rest form
 * the recent window. Cost is normalized by each window's duration to get a rate
 * (cost/second), so uneven window lengths compare fairly. A dimension is
 * flagged when its recent rate exceeds `factor` times its baseline rate.
 *
 * When `splitTime` is undefined, the midpoint of the observed timestamp range
 * is used. Records with a timestamp of 0 are treated as untimed and ignored.
 */
export function detectCreep(
  records: UsageRecord[],
  dimension: Dimension = "feature",
  splitTime?: number,
  factor = 2.0,
): Creep[] {
  const timed = records.filter((r) => (r.timestamp ?? 0) > 0);
  if (timed.length < 2) return [];

  const times = timed.map((r) => r.timestamp as number);
  const lo = Math.min(...times);
  const hi = Math.max(...times);
  if (hi === lo) return [];

  const split = splitTime ?? (lo + hi) / 2;
  const baseDur = Math.max(split - lo, 1e-9);
  const recentDur = Math.max(hi - split, 1e-9);

  const baseCost = new Map<string, number>();
  const recentCost = new Map<string, number>();
  for (const r of timed) {
    const key = keyOf(r, dimension);
    const bucket = (r.timestamp as number) < split ? baseCost : recentCost;
    bucket.set(key, (bucket.get(key) ?? 0) + recordCost(r));
  }

  const out: Creep[] = [];
  for (const [key, rc] of recentCost) {
    const baseRate = (baseCost.get(key) ?? 0) / baseDur;
    const recentRate = rc / recentDur;
    if (baseRate <= 0) continue; // no baseline to compare against; new, not creep
    if (recentRate > factor * baseRate) {
      out.push({
        key,
        baselineRate: Math.round(baseRate * 1e9) / 1e9,
        recentRate: Math.round(recentRate * 1e9) / 1e9,
        factor: Math.round((recentRate / baseRate) * 100) / 100,
      });
    }
  }
  return out.sort((a, b) => b.factor - a.factor);
}

/** Aggregates records and evaluates budget + anomalies in one pass. */
export function buildReport(
  records: UsageRecord[],
  dimension: Dimension = "feature",
  budget?: number,
  anomalyFactor = 3.0,
): Report {
  const stats = aggregate(records, dimension);
  const values = Object.values(stats);
  const totalCost = round6(values.reduce((acc, s) => acc + s.cost, 0));
  return {
    totalCost,
    totalCalls: values.reduce((acc, s) => acc + s.calls, 0),
    byDimension: stats,
    budgetExceeded: budget !== undefined && totalCost > budget,
    anomalies: detectAnomalies(stats, anomalyFactor),
  };
}
